hack/verify-licenses.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+if [[ "${TRACE-0}" == "1" ]]; then
+    set -o xtrace
+fi
+
+VERSION=${1}
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
+source "${REPO_ROOT}/hack/ensure-trivy.sh"
+
+TRIVY="${REPO_ROOT}/hack/tools/bin/trivy/${VERSION}/trivy"
+
+$TRIVY filesystem . --license-full --ignored-licenses MPL-2.0  --scanners license --exit-code 1 --severity MEDIUM,HIGH,CRITICAL
```

Comment:
More on license scanning with trivy here: https://aquasecurity.github.io/trivy/v0.44/docs/scanner/license/

I'm ignoring the MPL 2.0 license as we have a number of dependencies with that license. If  we want to more a more advanced config we could create one based on the CNCF license and allowlist.

One disadvantage of the Trivvy scan is that it seems to only scan dependencies in the `go.mod` indirect dependencies from `go.sum` are not scanned it looks like.

Reply:
This might be intentional. Not everything in go.sum actually makes it into the final binary

Reply:
> ideally, we should converge on the same tool used by K8s later on
I think k/k is using FOSSA which requires an API key to run so would be a higher maintenance burden. I'm not sure if it's much better than trivvy. 

What I meant by the above is that we can configure Trivy to be specific to the CNCF list of allowed licenses and the granted exceptions. There is some maintenance in this too as we'd have to keep the lists in sync.

Reply:
To be honest I'm struggling a bit between merging something as soon as we can in CAPI vs. exploring a bit more what k/k is doing, comparing it with what we do here and then decide what we should do in CAPI

But maybe I'm just lacking an overview

Reply:
>One disadvantage of the Trivvy scan is that it seems to only scan dependencies in the go.mod indirect dependencies from go.sum are not scanned it looks like.

This seems to be aligned to k/k?

> I'm ignoring the MPL 2.0 license as we have a number of dependencies with that license

Also in go.mod or only go.sum?

Reply:
Sounds good, let's do a bit more research and then discuss.

Thx for all the research you're doing on this.

One additional thing I'm struggling with is allowing MPL licences in general while it is not on the allow-list of CNCF. I think more information will help to make a call

Reply:
For the immediate feature we can be just very conservative with merging anything that bumps a hashicorp dependency or introduces a new dependency

Reply:
> Also in go.mod or only go.sum?

We have one MPL licensed projects in our go.mod after the recent changes:
`github.com/hashicorp/hcl` which is brought in by our usage of the viper library.

Reply:
> We have one MPL licensed projects in our go.mod after the recent changes:
github.com/hashicorp/hcl which is brought in by our usage of the viper library.

Let's see https://github.com/spf13/viper/issues/1608 :)